File: client/sconfd.py
# sconfd.py
# Author: https://github.com/hanneslehmann
# Licence: free to use, no warranties!

import sys
import redis
import time
from datetime import datetime

FORMAT = '%Y%m%d%H%M%S'

def dumpdata(obj,sep, f):
    if type(obj) == dict:
        for k, v in obj.items():
            if hasattr(v, '__iter__'):
                print k
                dumpdata(v)
            else:
#                print '%s %s %s' % (k, sep, v)
                f.write(k + sep + v + '\n')
                f.flush()
    elif type(obj) == list:
        for v in obj:
            if hasattr(v, '__iter__'):
                dumpdata(v)
            else:
                print v
    else:
        print obj

myfiles = []
whatchlist = set([])

def main(argv):
  r = redis.StrictRedis(host='localhost', port=6379, db=1)
  c = redis.StrictRedis(host='localhost', port=6379, db=1)
  lookfor = 'config:'+str(argv[0])+':'
  # print "my keys: " + lookfor
  # first get a list of config/template files I am interested in
  print "Watching..."
  for val in c.keys(lookfor+"*"):
    k = val.split(':')
    # myfiles.append(k[2])
    whatchlist.add(k[0]+":"+k[1])
    whatchlist.add("template:"+k[2])
    #print k[0]+":"+k[1]+":"+k[2]
    #print "template:"+k[2]
  # output to confirm we are whatching
  print whatchlist
  # now subscribing to changes
  pubsub = r.pubsub()
  pubsub.psubscribe('__keyevent@1__:hset')
  # when change happened we need to check if change
  # was done on a file we are watching
  for msg in pubsub.listen():
    s = str(msg['data'])
    params = s.split(":")
    ck = ""
    # check if the key contains at least lvl 2
    if len(params)>=2:
      ck = params[0]+":"+ params[1]
    # check if the key is being whatched
    if ck in whatchlist:
      # print "I was watching you!"
      key_index = 0
      if params[0] == 'config':
        key_index = 2
      if params[0] == "template":
        key_index = 1     
      if key_index > 0:
        # construct the keys
        key_c = lookfor+params[key_index]+':content'
        key_m = lookfor+params[key_index]+':meta'
        key_t = "template:"+params[key_index]
        # get data from template
        sep = c.hget(key_t + ":meta",'seperator')
        com = c.hget(key_t + ":meta",'comment')
        data_template = c.hgetall(key_t + ":content")
        # get data from client config
        fp  = c.hget(key_m,'filepath')
        data_client = c.hgetall(key_c)
        # copy template to temporary location
        data = data_template.copy()
        data.update(data_client)
        print "writing change of "+ ck +" to file: " + fp
        f = open(fp, 'w')
        f.write(com + ' Updated by sconfd at: ' + datetime.now().strftime(FORMAT) + '\n')
        f.flush()
        dumpdata (data, sep, f)
        f.close

if __name__ == "__main__":
    main(sys.argv[1:])

File: client/sconfd.go
/* sconfd.go
   Author: https://github.com/hanneslehmann
   Licence: free to use, no warranties!
*/
package main

import (
	"bufio"
	"flag"
	"fmt"
	"log"
	"os"
	"strings"
	"time"

	"github.com/garyburd/redigo/redis"
)

const (
	ADDRESS = "127.0.0.1:6379"
)

var (
	redis_address     string
	c, r              redis.Conn
	whatchlist        []string
	template_settings map[string]string
	config_settings   map[string]string
)

var Usage = func() {
	fmt.Fprintf(os.Stderr, "\n")
	fmt.Fprintf(os.Stderr, "Usage of sconfd:\n")
	flag.PrintDefaults()
	fmt.Fprintf(os.Stderr, "\n")
}

// Connect to Redis
func init_redis(url string) redis.Conn {
	var err error
	var c redis.Conn
	c, err = redis.Dial("tcp", url)
	if err != nil {
		log.Fatalf("Couldn't connect to Redis: %v\n", err)
	}
	return c
}

func contains(slice []string, item string) bool {
	set := make(map[string]struct{}, len(slice))
	for _, s := range slice {
		set[s] = struct{}{}
	}
	_, ok := set[item]
	return ok
}

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func main() {
	prefix := "template:"
	config := "config"

	var clientID string
	flag.StringVar(&clientID, "id", "", "the unique Client ID for identification (mandatory)")
	flag.StringVar(&redis_address, "a", ADDRESS, "the URL with format IP:PORT for connecting database (optional)")
	flag.Parse()

	c := init_redis(redis_address)
	defer c.Close()
	r := init_redis(redis_address)
	defer r.Close()

	c.Do("SELECT", 1)
	r.Do("SELECT", 1)

	if len(clientID) > 1 {
		fmt.Printf("Connected to: tcp://%s \n", redis_address)
		config = fmt.Sprint("config:", clientID, ":")
		fmt.Printf("Looking for changes of config:%s and ", clientID)
		strs, err := redis.Strings(c.Do("KEYS", fmt.Sprint(config, "*")))
		whatchlist := []string{config}
		if err != nil {
			log.Fatal(err)
		}
		for _, element := range strs {
			arr := strings.Split(element, ":")
			if !contains(whatchlist, fmt.Sprint("template:", arr[2], ":")) {
				whatchlist = append(whatchlist, fmt.Sprint("template:", arr[2], ":"))
				fmt.Printf(" template:%s", arr[2])
			}
		}
		fmt.Printf(" \n")
		psc := redis.PubSubConn{Conn: r}
		psc.Subscribe("__keyevent@1__:hset")
		for {
			switch n := psc.Receive().(type) {
			case redis.Message:
				params := strings.Split(string(n.Data), ":")
				if len(params) > 2 {
					ck := fmt.Sprint(params[0], ":", params[1], ":")
					if contains(whatchlist, ck) {
						key_index := 0
						switch params[0] {
						case "template":
							key_index = 1
						case "config":
							key_index = 2
						}
						if key_index > 0 {
							key_c := fmt.Sprint(config, params[key_index], ":content")
							key_m := fmt.Sprint(config, params[key_index], ":meta")
							key_tc := fmt.Sprint(prefix, params[key_index], ":content")
							key_tm := fmt.Sprint(prefix, params[key_index], ":meta")
							com, err := redis.String(c.Do("HGET", key_tm, "comment"))
							check(err)
							sep, err := redis.String(c.Do("HGET", key_tm, "seperator"))
							check(err)
							fp, err := redis.String(c.Do("HGET", key_m, "filepath"))
							check(err)
							template_settings, err := redis.StringMap(c.Do("HGETALL", key_tc))
							check(err)
							config_settings, err := redis.StringMap(c.Do("HGETALL", key_c))
							check(err)
							for key, value := range config_settings {
								_, ok := template_settings[key]
								if ok {
									template_settings[key] = value
								}
							}
							f, err := os.Create(fp)
							check(err)
							defer f.Close()
							w := bufio.NewWriter(f)
							w.WriteString(fmt.Sprint(com, " ", time.Now().Format(time.RFC850), " Update from sconfd\n"))
							for key, value := range template_settings {
								w.WriteString(fmt.Sprint(key, " ", sep, " ", value, "\n"))
							}
							w.Flush()
							f.Close()
							fmt.Printf("File %s updated with changes from %s\n", fp, params[0])
						}
					}
				}
			case error:
				fmt.Printf("error: %v\n", n)
				return
			}
		}

	} else {
		Usage()
	}

}
